### PR TITLE
Add proxy server feature.

### PR DESCRIPTION
**Pull request by Darkclainer**
Now in configuration in "proxy" section you can specify two new options:
1. url - is a proxy url
2. mode - it is string, one of "all", "missing", "none"

Issue #33

### CODE REVIEW

**Comment by aperezg** [comment]
Can you check  it, @joanlopez?

File: internal/config.go [hunk 1]
```diff
@@ -26,6 +28,43 @@ type ConfigCORS struct {
...
 }
 
+// ConfigProxy is a representation of section proxy of the yaml
+type ConfigProxy struct {
+	Url  string    `yaml:"url"`
+	Mode ProxyMode `yaml:"mode"`
+}
+
+// ProxyMode is enumeration of proxy server modes
+type ProxyMode int
```
Comment (aperezg) [inline internal/config.go]:
```suggestion
type ProxyMode uint8
```

File: internal/config.go [hunk 2]
```diff
@@ -26,6 +28,43 @@ type ConfigCORS struct {
...
+	// ProxyNone server is off
+	ProxyNone ProxyMode = iota
+	// ProxyMissing handle only missing requests are proxied
+	ProxyMissing
+	// ProxyAll all requests are proxied
+	ProxyAll
+)
+
+// UnmarshalYAML implementation of yaml.Unmarshaler interface
+func (mode *ProxyMode) UnmarshalYAML(unmarshal func(interface{}) error) error {
```
Comment (aperezg) [inline internal/config.go]: you cand add somthing like this:

```suggestion
func proxyModeParseString(t string) (ProxyMode, error) {
	m := map[string]ProxyMode{
		"none":    ProxyNone,
		"missing": ProxyMissing,
		"all":     ProxyAll,
	}

	p, ok := m[t]
	if !ok {
		return ProxyNone, fmt.Errorf("unknown proxy mode: %s", t)
	}

	return p, nil
}
```

with this suggestion to could avoid the switch and only do 
```go
...
return proxyModeParseString(textMode)
```

Comment (Darkclainer) [reply internal/config.go]: What are the cons in using map here? Maybe it is a bit shorter/

File: internal/config.go (hunk 1 above)
Comment (Darkclainer) [reply internal/config.go]: Why do we need specifically uint8?

Comment (aperezg) [reply internal/config.go]: we prefer declare the enums with a more accurate type possible as standard, 8bits is more than enough to a enum and in most case u don't need negative numbers, so `uint8` is a elegant solution

File: internal/config.go (hunk 2 above)
Comment (aperezg) [reply internal/config.go]: Sorry, I don't understand your question... In your commit I don't see any map, the map is my suggestion, could you explain me a little bit more? thanks :)

File: internal/server/http/proxy.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		return nil, err
+	}
+	reverseProxy := httputil.NewSingleHostReverseProxy(u)
+	return &Proxy{reverseProxy, mode}, nil
+}
+
+// Handler returns handler that sends request to another server.
+func (p *Proxy) Handler() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		p.server.ServeHTTP(w, r)
```
Comment (aperezg) [inline internal/server/http/proxy.go]: you need to update the headers to allow for SSL redirection:

```go
                r.URL.Host = p.url.Host
		r.URL.Scheme = p.url.Scheme
		r.Header.Set("X-Forwarded-Host", r.Header.Get("Host"))
		r.Host = p.url.Host
```

For that you will need to persist the result of url.Parse into the proxy struct.

**Comment by aperezg** [comment]
All options that we include on the config file, usually are on the command line also. 

I think that you could add something like that:
`-proxy-url="https://swapi.co" -proxy-mode=missing`

And if the `proxy-mode` is not declare using by default none, and if the `proxy_mode` != none and `proxy_url` is not defined get an early error and close the application.

File: internal/server/http/proxy.go (hunk 3 above)
Comment (aperezg) [reply internal/server/http/proxy.go]: I tested with the comment that I suggested and, all working very fine :)

**Comment by aperezg** [comment]
Could you update the README.md with how to use the proxy?

In the section **Using Killgrave**

Below to the sentence: "Or custome your server with this flags:" we're printing an output of the help command, so when you've added your new flags, you will need to update this with your output, as now.

And in the section "Use `killgrave` with config file:"

Modify the example with the proxy, i.e:

```yml
#config.yml

imposters_path: "imposters"
port: 3000
host: "localhost"
proxy:
  url: https://example.com
  mode: missing
cors:
  methods: ["GET"]
  headers: ["Content-Type"]
  exposed_headers: ["Cache-Control"]
  origins: ["*"]
  allow_credentials: true
```

And if you want you could include a senction like **#CORS** section, where you explain how the proxy works and which are the options to configure the proxy.

Thanks!

**Comment by aperezg** [comment]
Follow #44

**Pull request closed without merging**